vendor/github.com/pstibrany/services/basic_service.go

```diff
@@ -0,0 +1,341 @@
+type StoppingFn func() error
+
+// BasicService implements contract of Service interface, using three supplied functions: StartingFn, RunningFn and StoppingFn.
+// When service is started, these three functions are called as service transitions to Starting, Running and Stopping state.
+//
+// Since they are called sequentially, they don't need to synchronize access on the state.
+// (In other words: StartingFn happens-before RunningFn, RunningFn happens-before StoppingFn).
+//
+// All three functions are called at most once. If they are nil, they are not called and service transitions to the next state.
+//
+// Context passed to StartingFn and RunningFn function is canceled when StopAsync() is called, or service enters Stopping state.
+// This context can be used to start additional tasks from inside StartingFn or RunningFn.
+// Same context is available via ServiceContext() method (not part of Service interface).
+//
+// Possible orders of how functions are called:
+//
+// * 1. StartingFn -- if StartingFn returns error, no other functions are called.
+//
+// * 1. StartingFn, 2. StoppingFn -- StartingFn doesn't return error, but StopAsync is called while running
+// StartingFn, or context is canceled from outside while StartingFn still runs.
+//
+// * 1. StartingFn, 2. RunningFn, 3. StoppingFn -- this is most common, when StartingFn doesn't return error,
+// service is not stopped and context isn't stopped externally while running StartingFn.
+type BasicService struct {
+	// functions only run, if they are not nil. If functions are nil, service will effectively do nothing
+	// in given state, and go to the next one without any error.
+	startFn    StartingFn
+	runningFn  RunningFn
+	stoppingFn StoppingFn
+
+	// everything below is protected by this mutex
+	stateMu     sync.RWMutex
+	state       State
+	failureCase error
+	listeners   []chan func(l Listener)
+
+	// closed when state reaches Running, Terminated or Failed state
+	runningWaitersCh chan struct{}
+	// closed when state reaches Terminated or Failed state
+	terminatedWaitersCh chan struct{}
+
+	serviceContext context.Context
+	serviceCancel  context.CancelFunc
+}
+
+func invalidServiceStateError(state, expected State) error {
+	return fmt.Errorf("invalid service state: %v, expected: %v", state, expected)
+}
+
+func invalidServiceStateWithFailureError(state, expected State, failure error) error {
+	return fmt.Errorf("invalid service state: %v, expected: %v, failure: %w", state, expected, failure)
+}
+
+// Returns service built from three functions (using BasicService).
+func NewService(start StartingFn, run RunningFn, stop StoppingFn) Service {
+	bs := &BasicService{}
+	InitBasicService(bs, start, run, stop)
+	return bs
+}
+
+// Initializes basic service. Should only be called once. This method is useful when
+// BasicService is embedded as part of bigger service structure.
+func InitBasicService(b *BasicService, start StartingFn, run RunningFn, stop StoppingFn) {
```

Comment:
My idea was that when BasicService is embedded as part of bigger service, then this may be needed, but I think you are right and `NewBasicService` should be enough. Embedding then needs to use a pointer, to be able to use `NewBasicService`, but that's ok. I'll give it a try.

Reply:
I've now removed `Init*` functions, and it's nicer this way. Thanks.